package com.dev.eventmanager.web;

import jakarta.persistence.EntityNotFoundException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

import java.time.OffsetDateTime;
import java.util.stream.Collectors;

@ControllerAdvice
public class GlobalExceptionHandler {

    private static final Logger log = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    @ExceptionHandler({
            MethodArgumentNotValidException.class,
            IllegalArgumentException.class,
            IllegalStateException.class
    })
    public ResponseEntity<ServerErrorDto> handleValidationException(
            Exception e
    ) {
        log.error("Got validation exception", e);


        String detailedMessage = e instanceof MethodArgumentNotValidException
                ? constractMethodArgumentNotValidMethod((MethodArgumentNotValidException) e)
                : e.getMessage();

        var errorDto = new ServerErrorDto(
                "Некорректный запрос",
                detailedMessage,
                OffsetDateTime.now());
        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(errorDto);

    }

    @ExceptionHandler
    public ResponseEntity<ServerErrorDto> handleGenerisException(
            Exception e
    ) {
        if (e instanceof org.springframework.security.core.AuthenticationException) {
            throw (org.springframework.security.core.AuthenticationException) e;
        }

        log.error("Server error", e);
        var errorDto = new ServerErrorDto(
                "Внутренняя ошибка на сервере",
                e.getMessage(),
                OffsetDateTime.now());
        return ResponseEntity
                .status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(errorDto);

    }

    @ExceptionHandler(EntityNotFoundException.class)
    public ResponseEntity<ServerErrorDto> handleNotFoundException(
            EntityNotFoundException e
    ) {
        log.error("Got no such element exception", e);
        var errorDto = new ServerErrorDto(
                "Сущность не найдена",
                e.getMessage(),
                OffsetDateTime.now());
        return ResponseEntity
                .status(HttpStatus.NOT_FOUND)
                .body(errorDto);

    }

    private String constractMethodArgumentNotValidMethod(MethodArgumentNotValidException e) {

        return e.getBindingResult()
                .getFieldErrors()
                .stream()
                .map(error -> error.getField() + ": " + error.getDefaultMessage())
                .collect(Collectors.joining(", "));
    }
    @ExceptionHandler(AccessDeniedException.class)
    public ResponseEntity<ServerErrorDto> handleAccessDenied(AccessDeniedException ex) {
        ServerErrorDto error = new ServerErrorDto(
                "Forbidden",
                ex.getMessage(),
                OffsetDateTime.now()
        );
        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(error);
    }

}
